src/scarabee/_scarabee/cmfd.cpp

```diff
@@ -325,26 +460,758 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
+  }
+  keff_ = keff;
 }
 
-void CMFD::solve(MOCDriver& moc) {
-  this->normalize_currents();
-  this->compute_homogenized_xs_and_flux(moc);
+void CMFD::fixed_source_solve() {
+  // Solve fixed source problem
+  // Subtract fission source from loss matrix
+  Eigen::SparseMatrix<double> L = M_ - QM_;
+
+  Eigen::VectorXd new_flux(ng_ * nx_ * ny_);
+
+  // Create a solver for the problem
+  Eigen::BiCGSTAB<Eigen::SparseMatrix<double>> solver;
+  solver.compute(L);
+  solver.setTolerance(1.E-10);
+
+  if (solver.info() != Eigen::Success) {
+    std::stringstream mssg;
+    mssg << "Could not initialize CMFD iterative solver";
+    spdlog::error(mssg.str());
+    throw ScarabeeException(mssg.str());
+  }
+
+  // Get new flux
+  new_flux = solver.solveWithGuess(extern_src_, flux_cmfd_);
+  // For some reason, this doesn't seem to be working with the new versions
+  // of Eigen, despite clearly succeeding. Just commenting it out for now.
+  // if (solver.info() != Eigen::Success) {
+  //   spdlog::error("Solution impossible.");
+  //   throw ScarabeeException("Solution impossible");
+  // }
+
+  flux_cmfd_ = new_flux;
 }
 
-}  // namespace scarabee
\ No newline at end of file
+void CMFD::update_moc_fluxes(MOCDriver& moc) {
+  // Update MOC FSR scalar fluxes
+  const std::size_t tot_cells = ny_ * nx_;
+  bool flux_update_warning = false;
+
+  // Normalize homogenized moc flux and cmfd flux
+  flux_ /= xt::sum(flux_)();
+  flux_cmfd_ /= flux_cmfd_.sum();
+
+  // Precompute update ratio in each CMFD cell
+  std::size_t i = 0;
+  std::size_t j = 0;
+  for (std::size_t l = 0; l < tot_cells; l++) {
+    for (std::size_t g = 0; g < moc_to_cmfd_group_map_.size(); g++) {
+      const std::size_t G = moc_to_cmfd_group_map_[g];
+      const std::size_t linear_indx = G * tot_cells + l;
+      const double invs_flx = 1. / flux_(G, i, j);
+      double ratio = flux_cmfd_(linear_indx) * invs_flx;
+      update_ratios_(linear_indx) = ratio;
+      // Don't warn on first moc iteration
+      if (ratio > 20.0 && moc_iteration_ > 1) {
+        flux_update_warning = true;
+      }
+      // Check that the update ratio is valid
+      if (ratio < 0.0) {
+        auto mssg =
+            "Negative CMFD flux update ratio. Try using Larsen correction";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
+      if (ratio == 0.0){
+        auto mssg =
+            "CMFD flux update is zero.";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
+      if (std::isinf(ratio)) {
+        auto mssg = "CMFD flux update ratio is inf";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
+      if (std::isnan(ratio)) {
+        auto mssg = "CMFD flux update ratio is NaN";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
+      // Clamp CMFD flux ratio after a certain # of solves
+      if (cmfd_solves_ > unbounded_cmfd_solves_) {
+        std::clamp(ratio, 0.05, 20.0);
+      }
```

Comment:
Here, you are clamping the update ratio, but you never actually clamp / update the value in the update_ratios_ array.